dae/dae/task_graph/executor.py

```diff
@@ -0,0 +1,239 @@
+from abc import abstractmethod
+from copy import copy
+from typing import Any, Iterator
+import logging
+
+from dae.task_graph.graph import TaskGraph, Task
+from dae.task_graph.cache import TaskCache, NoTaskCache, CacheRecordType
+
+
+logger = logging.getLogger(__file__)
+
+
+class TaskGraphExecutor:
+    """Class that executes a task graph."""
+
+    @abstractmethod
+    def execute(self, task_graph: TaskGraph) -> Iterator[tuple[Task, Any]]:
+        """Start executing the graph.
+
+        Return an iterator that yields the task in the graph as they finish not
+        nessessarily in DFS or BFS order.
+        """
+
+    @abstractmethod
+    def get_active_tasks(self) -> list[Task]:
+        """Return the list of tasks currently being processed."""
+
+
+class AbstractTaskGraphExecutor(TaskGraphExecutor):
+    """Executor that walks the graph in order that satisfies dependancies."""
+
+    def __init__(self, task_cache: TaskCache = NoTaskCache()):
+        super().__init__()
+        self._task_cache = task_cache
+        self._executing = False
+
+    def execute(self, task_graph: TaskGraph) -> Iterator[tuple[Task, Any]]:
+        assert not self._executing, \
+            "Cannot execute a new graph while an old one is still running."
+        self._check_for_cyclic_deps(task_graph)
+
+        self._executing = True
+        self._task_cache.set_task_graph(task_graph)
+        already_computed_tasks = {}
+        for task_node in self._in_exec_order(task_graph):
+            record = self._task_cache.get_record(task_node)
+            if record.type == CacheRecordType.COMPUTED:
+                already_computed_tasks[task_node] = record.result
+                self._set_task_result(task_node, record.result)
+            else:
+                self._queue_task(task_node)
+        return self.__await_tasks(already_computed_tasks)
+
+    def __await_tasks(self, already_computed_tasks) \
+            -> Iterator[tuple[Task, Any]]:
```

Comment:
This name is too similar to `_await_tasks,` and it was confusing at first.

Reply:
@lchorbadjiev Fixed.